src/internal/components/file-input/index.tsx

```diff
@@ -0,0 +1,126 @@
+import { useFormFieldContext } from '../../../contexts/form-field';
+import ScreenreaderOnly from '../../components/screenreader-only';
+import { fireNonCancelableEvent } from '../../events';
+import checkControlled from '../../hooks/check-controlled';
+import useForwardFocus from '../../hooks/forward-focus';
+import { useUniqueId } from '../../hooks/use-unique-id';
+import { joinStrings } from '../../utils/strings';
+import { FileInputProps } from './interfaces';
+
+import styles from './styles.css.js';
+
+export { FileInputProps };
+
+const InternalFileInput = React.forwardRef(
+  (
+    {
+      accept,
+      ariaRequired,
+      ariaLabel,
+      multiple = false,
+      value,
+      onChange,
+      variant = 'button',
+      children,
+      ...restProps
+    }: FileInputProps,
+    ref: Ref<FileInputProps.Ref>
+  ) => {
+    const uploadInputRef = useRef<HTMLInputElement>(null);
+    const uploadButtonLabelId = useUniqueId('upload-button-label');
+    const formFieldContext = useFormFieldContext(restProps);
+    const selfControlId = useUniqueId('upload-input');
+    const controlId = formFieldContext.controlId ?? selfControlId;
+
+    useForwardFocus(ref, uploadInputRef);
+
+    const [isFocused, setIsFocused] = useState(false);
+    const onUploadButtonClick = () => uploadInputRef.current?.click();
+    const onUploadInputFocus = () => setIsFocused(true);
+    const onUploadInputBlur = () => setIsFocused(false);
+
+    const onUploadInputChange = ({ target }: ChangeEvent<HTMLInputElement>) => {
+      fireNonCancelableEvent(onChange, { value: target.files ? Array.from(target.files) : [] });
+    };
+
+    checkControlled('FileInput', 'value', value, 'onChange', onChange);
+
+    const nativeAttributes: React.HTMLAttributes<HTMLInputElement> = {
+      'aria-label': ariaLabel || children,
+      'aria-labelledby': joinStrings(formFieldContext.ariaLabelledby, uploadButtonLabelId),
+      'aria-describedby': formFieldContext.ariaDescribedby,
+    };
+    if (formFieldContext.invalid) {
+      nativeAttributes['aria-invalid'] = true;
+    }
+    if (ariaRequired) {
+      nativeAttributes['aria-required'] = true;
+    }
+
+    if (variant === 'icon' && !ariaLabel) {
+      warnOnce('FileInput', 'Aria label is required with icon variant.');
+    }
+
+    // Synchronizing component's value with the native file input state.
+    useEffect(() => {
+      /* istanbul ignore next: The DataTransfer is not available in jsdom. */
+      if (window.DataTransfer) {
+        const dataTransfer = new DataTransfer();
+        for (const file of value) {
+          dataTransfer.items.add(file);
+        }
+        uploadInputRef.current!.files = dataTransfer.files;
+      }
+      if (uploadInputRef.current) {
+        uploadInputRef.current.value = ''; // reset value to allow calling onChange when the same file is uploaded again
+      }
+    }, [value]);
+
+    return (
+      <div className={clsx(styles['file-input-container'])}>
```

Comment:
Minor: Why not root but file-input-container?

Reply:
It's just what it was called previously. I'll update to be `root` for consistency.